File: ios/Sources/HydraPayments/Exceptions.swift
import Foundation

public class HydraError: Error, CustomStringConvertible {
    public let statusCode: Int
    public let errorCode: String
    public let message: String
    public let details: [String: String]?

    public init(message: String, statusCode: Int = 500, errorCode: String = "API_ERROR", details: [String: String]? = nil) {
        self.message = message
        self.statusCode = statusCode
        self.errorCode = errorCode
        self.details = details
    }

    public var description: String {
        "[\(errorCode)] \(message)"
    }
}

public final class AuthenticationError: HydraError {
    public init(message: String, details: [String: String]? = nil) {
        super.init(message: message, statusCode: 401, errorCode: "AUTHENTICATION_ERROR", details: details)
    }
}

public final class ValidationError: HydraError {
    public init(message: String, details: [String: String]? = nil) {
        super.init(message: message, statusCode: 400, errorCode: "VALIDATION_ERROR", details: details)
    }
}

public final class NotFoundError: HydraError {
    public init(message: String, details: [String: String]? = nil) {
        super.init(message: message, statusCode: 404, errorCode: "NOT_FOUND", details: details)
    }
}

File: ruby/lib/hydra_payments/exceptions.rb
module HydraPayments
  class HydraError < StandardError
    attr_reader :error_code, :status_code, :details

    def initialize(message, status_code:, error_code: 'API_ERROR', details: nil)
      super(message)
      @error_code = error_code
      @status_code = status_code
      @details = details
    end
  end

  class AuthenticationError < HydraError
    def initialize(message, details: nil)
      super(message, status_code: 401, error_code: 'AUTHENTICATION_ERROR', details: details)
    end
  end

  class ValidationError < HydraError
    def initialize(message, details: nil)
      super(message, status_code: 400, error_code: 'VALIDATION_ERROR', details: details)
    end
  end

  class NotFoundError < HydraError
    def initialize(message, details: nil)
      super(message, status_code: 404, error_code: 'NOT_FOUND', details: details)
    end
  end
end
